apis/apim/v1alpha1/api_types.go

```diff
@@ -33,33 +33,11 @@ type API_Metadata struct {
 	OpenAPIRef  OPARef `json:"openAPIRef,omitempty"`
 }
 
-type Operation_GET struct{}
-
-type Upstream struct {
-	URL         string `json:"url,omitempty"`
-	ServiceName string `json:"serviceName,omitempty"`
-	Namespace   string `json:"namespace,omitempty"`
-	Port        uint32 `json:"port,omitempty"`
-	TLS         string `json:"tls,omitempty"`
-}
-
-type Operation_POST struct {
-	Upstreams []Upstream `json:"upstreams"`
-}
-
-type Operation struct {
-	Get  *Operation_GET  `json:"get,omitempty"`
-	Post *Operation_POST `json:"post,omitempty"`
-}
-
 type APISpec struct {
-	Domains            []string             `json:"domains"`
-	Gateways           []string             `json:"gateways,omitempty"`
-	Info               API_Metadata         `json:"info"`
-	Upstreams          []Upstream           `json:"upstreams"`
-	Paths              map[string]Operation `json:"paths"`
-	AuthConfigSelector map[string]string    `json:"authConfigSelector,omitempty"`
-	RateLimitSelector  *map[string]string   `json:"rateLimitSelector,omitempty"`
+	Info                   API_Metadata      `json:"info"`
+	VirtualServiceSelector map[string]string `json:"virtualServiceSelector,omitempty"`
```

Comment:
I do not get this :)

Reply:
The customer will be the owner of the virtualservice objects? In that case,  kuadrant API object does no longer make any sense. Annotations and labels can be used to link VirtualServices, AuthConfigs, AuthortizationPolicies and RateLimitPolicies. 

IMO the current kuadrant API CRD is a good compromise for an API owner between capabilities and complexity. It is made for API owners, in their own language and terms.